crates/object_store_util/src/shared.rs

```diff
@@ -86,14 +89,29 @@ impl ObjectStore for SharedObjectStore {
     }
 
     async fn copy_if_not_exists(&self, from: &Path, to: &Path) -> Result<()> {
-        self.inner.copy_if_not_exists(from, to).await
+        match self.inner.copy_if_not_exists(from, to).await {
+            Ok(_) => Ok(()),
+            Err(ObjectStoreError::NotSupported { .. }) => {
+                // Go with the poor man's copy-if-not-exists: try a regular copy if the path doesn't exist
+                match self.head(to).await {
+                    Ok(_) => {
+                        return Err(ObjectStoreError::AlreadyExists {
+                            path: to.to_string(),
+                            source: anyhow!(
+                            "Object at path {to} already exists, can't perform copy-if-not-exists"
+                        )
+                            .into(),
+                        })
+                    }
+                    Err(ObjectStoreError::NotFound { .. }) => self.copy(from, to).await,
+                    Err(e) => Err(e),
+                }
+            }
+            Err(e) => Err(e),
+        }
     }
```

Comment:
> What I think we can do is change lease initialization logic to:
> 
>     1. Read lease. If it exists:
> 
> 
>     * return Ok if process id written in lease matches leaser's process id
> 
>     * return Err if process ids don't match
> 
> 
>     2. Put new lease (puts are atomic in s3)
> 
>     3. Read lease again to ensure no other process wrote the lease between steps 1 & 2.

In the end I realized that this alone wasn't going to work, since apart from us using `rename_if_not_exists` during lease initialization, and more importantly, `delta-rs` also uses it when committing a transaction. So I went with the above as a way to resolve both.

In this way, `NativeTableStorage` can pass a working object store to delta-rs, which can in turn commit the transactions. On the other hand the metastore client also gets initialized with a `SharedObjectStore` now, so the lease initialization will also work.

However, this does mean that races are still a potential problem, arguably more pronounced for transaction consistency than lease initialization. Therefore, in the long term we might need to implement the S3 lock client after all (or use `deltalake::s3::S3StorageBackend` which has one built in). 

One might also argue that this shouldn't be too big of a problem, since writes shouldn't be as frequent as reads (for OLAP workloads at least). Nonetheless, we could add a warning statement or an additional CLI switch so as to make the user fully aware of potential issues.

Reply:
> Therefore, in the long term we might need to implement the S3 lock client after all (or use deltalake::s3::S3StorageBackend which has one built in).

is there a downside to using the deltalake client? does it require changing some kind of "on-disk" format? should we just open an issue for that?

Reply:
Not really, the only issue I see is the extra config needed to get the S3 lock working.

That said, I think I recall there were some plans of actually removing the delta-rs clients altogether: https://github.com/delta-io/delta-rs/issues/1521